README.md
# NAME

Data::Embed - embed arbitrary data in a file

# VERSION

This document describes Data::Embed version 0.32.

# SYNOPSIS

    use Data::Embed qw< embed embedded >;

    # this is the file where thing will be embedded, at the end
    my $container = '/path/to/some/file';

    # first of all we embed an external file
    my $datafile  = '/path/to/data.tar.gz';
    embed($container, name => 'data.tar.gz', filename => $datafile);

    # we can also embed some data, directly
    use Data::Dumper;
    my $conf = { ... };
    embed($container, name => 'config.yml', data => Dumper($conf));

    # if the data is in a scalar but it's huge, use filename and
    # pass a reference to the scalar so no copy will happen
    my $huge_png = ...;
    embed($container, name => 'image.png', filename => \$huge_png);

    # to retrieve the stuff, use embedded()
    my @files = embedded($container);

    # each item in @files is a Data::Embed::File object

    # get whole contents of file
    my $config_text = $files[1]->contents();

    # otherwise, you can get a filehandle and use it, e.g. to
    # dump it on standard output
    my $data_fh = $files[0]->fh();
    binmode STDOUT;
    print {*STDOUT} <$data_fh>;

    # or save the file back, using the available name
    open my $ofh, '>:raw', $file[2]->name(); # well, do your checks!
    my $ifh = $files[2]->fh();
    while (! eof $ifh) {
       read $ifh, my $buffer, 4096
          or last; # do proper checks in production!
       print {$ofh} $buffer;
    }

# DESCRIPTION

This module allows you to manage embedding data at the end of other
files, providing both means for embedding the data (["embed"](#embed) and
["writer"](#writer)) and accessing them (["embedded"](#embedded) and ["reader"](#reader)).

How can this be helpful? For example, suppose that you want to bring
some data along with your perl script, some of which might be binary
(e.g. an image, or a tar archive), you can embed these data inside the
perl and then retrieve them. For example, this can be the basis for an
installer.

For embedding data, you can use the ["embed"](#embed) function. See the relevant
documentation or the examples in the ["SYNOPSYS"](#synopsys) to use it properly.

For extracting the embedded data, you can use the ["embedded"](#embedded) function
and access each embedded file as a [Data::Embed::File](https://metacpan.org/pod/Data::Embed::File) object. You can
then use its methods `contents` for accessing the whole data, or get a
filehandle through `fh` and avoid getting the whole data in memory at
once.

Note: the filehandle provided by the `fh` method of
[Data::Embed::File](https://metacpan.org/pod/Data::Embed::File) is actually a [IO::Slice](https://metacpan.org/pod/IO::Slice) object, so it might not
support all the functions/methods of a regular filehandle.

You can also access the lower level interface through the two functions
["reader"](#reader) and ["writer"](#writer). See the documentation for
[Data::Embed::Reader](https://metacpan.org/pod/Data::Embed::Reader) and [Data::Embed::Writer](https://metacpan.org/pod/Data::Embed::Writer).

# FUNCTIONS

## **embed**

    embed($hashref); # OR
    embed(%keyvalue_pairs);

Embed new data inside a container file.

Parameters can be passed as key-value pairs either directly or through a
hash reference. The following keys are supported:

- `container`

    shortcut to specifying the same input and output, i.e. the value will be
    replicated both on the `input` and `output` keys below. Caller still
    has to ensure that the two are compatible. Provision of a filehandle is
    currently not supported.

- `input`

    any previous container file to use as base for the generated container.
    If missing, no previous data will be considered (like starting from an
    empty file). Can be:

    - the `-` string in a plain scalar, in which case standard input is
    considered
    - any other string in a plain scalar, considered to be a file name
    - a plain reference to a scalar, considered to hold the input data
    - something that supports the filehandle interface for reading

- `output`

    the target container for the newly generated archive. Might be the same
    as the input or different; in the latter case, the input will be copied
    over the output, apart from the bits regarding the management of the
    inclusions. Can be:

    - missing/undefined or the `-` string in a plain scalar, in which case
    standard output is used
    - any other string in a plain scalar, considered to be a file name
    - a plain reference to a scalar, considered to be the target scalar to
    hold the data
    - something that supports the filehandle interface for printing. You
    should not provide the same filehandle for both input and output, even
    if you opened it in read-write mode. This limitation might be removed in
    the future.

- `name`

    the name to associate to the section, optionally. If missing it will be
    set to the empty string

- `fh`

    the filehandle from where data should be taken. The filehandle will be
    exausted starting from its current position

- `filename`

    a filename or a reference to a scalar where data will be read from

- `data`

    a scalar from where data will be read. If you have a huge amount of
    data, it's better to use the `filename` key above passing a reference
    to the scalar holding the data.

Options `fh`, `filename` and `data` are exclusive and will be
considered in the order above (first come, first served).

This function does not return anything.

## **embedded**

Get a list of the embedded files inside a target container. The calling
syntax is as follows:

    my $arrayref = embedded($container); # scalar context, OR
    my @files    = embedded($container); # list context

The only input parameter is the `$container` to use as input. It can be
either a real filename, or a filehandle.

Depending on the context, a list will be returned (in list context) or
an array reference holding the list.

Whatever the context, each item in the list is a [Data::Embed::File](https://metacpan.org/pod/Data::Embed::File)
object that you can use to access the embedded file data (most notably,
you'll be probably using its `contents` or `fh` methods).

## **generate\_module\_from\_file**

    # when %args includes details for an output channel
    generate_module_from_file(%args);

    # in case no output is provided in %args:
    my $text = generate_module_from_file(%args);

Generate a module's file contents from a file. The module contains code
of a package that has code to read the included data. Arguments are:

- package

    the name of the package that will be put into the module. This is a
    mandatory parameter.

- output

    the output channel. If not present, the output will be provided as a
    string returned by the function, otherwise you can provide

    - a filehandle where the output will be printed
    - a reference to a scalar (it will be filled with the contents)
    - the `-` string, in which case the output will be printed
    to STDOUT
    - a filename

- output\_from\_package

    if this key is present and true, the `output` parameters is overridden
    and generated automatically from the package name provided in key
    `package`. The generated file will assume that the file is contained in
    the _normal_ path under a `lib` directory, e.g. if the package name is
    `Some::Module` then the generated filename will be
    `lib/Some/Module.pm`.

- fh

    a filehandle where data will be read from

- filename

    the input will be taken from the provided filename

- dataref

    the input will be taken from the scalar pointed by the
    reference

- data

    the input is taken from the scalar provided with the data key

Input keys are `fh`, `filename`, `dataref` and `data`. In case
multiple of them are present, they will be considered in the order
specified.

## **reader**

This is a convenience wrapper around the constructor for
[Data::Embed::Reader](https://metacpan.org/pod/Data::Embed::Reader).

## **reassemble**

    # when %args includes details for an output channel
    reassemble(%args);

Reassemble a target container fitting new input sequence. The available
arguments are:

- sequence

    the sequence of items that have to be embedded. Each item can be:

    - a [Data::Embed::File](https://metacpan.org/pod/Data::Embed::File) (e.g. coming from what you read from some other
    file)
    - a reference to a hash whose contents is compatible with what expected by
    [Data::Embed::Writer::add](https://metacpan.org/pod/Data::Embed::Writer::add).

- target

    the target container. It can be:

    - a _filehandle_
    - a _filename_ (including `-`, that does what you mean)
    - a _reference to a scalar_

    If the file or reference to a scalar are used, it will make sure to
    avoid clobbering. In particular, the _prefix_ data (i.e. data that is
    not part of the list of files) will be preserved.

## **writer**

This is a convenience wrapper around the constructor for
[Data::Embed::Writer](https://metacpan.org/pod/Data::Embed::Writer).

# BUGS AND LIMITATIONS

Report bugs either through RT or GitHub (patches welcome).

Passing the same filehandle for both `input` and `output` in ["embed"](#embed)
is not supported. This applies to `container` too.

# SEE ALSO

[Data::Section](https://metacpan.org/pod/Data::Section) covers a somehow similar need but differently. In
particular, you should look at it if you want to be able to modify the
data you want to embed directly, e.g. if you are embedding some textual
templates that you want to tweak.

# AUTHOR

Flavio Poletti <polettix@cpan.org>

# COPYRIGHT AND LICENSE

Copyright (C) 2014-2016 by Flavio Poletti <polettix@cpan.org>

This module is free software. You can redistribute it and/or modify it
under the terms of the Artistic License 2.0.

This program is distributed in the hope that it will be useful, but
without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.


Changes
{{$NEXT}}

0.32  2016-04-19 06:46:20 CEST
   - release after 299 PASS on CPAN Testers (thanks!)

0.31_1  2016-02-06 13:28:21 CET
   - fixed bug occurring when calling prefix before files
   - worked around bug in pre-5.14 perls (open() on scalar refs)

0.31  2016-02-05 15:25:06 CET
   - removed duplicate test file pod.t
   - set version for all modules to make indexer happy

0.3_02  2016-02-05 09:58:57 CET
   - added IO::Slice to the requirements

0.3_01  2016-02-05 07:20:00 CET
   - added cpanfile for dependencies

0.30  2016-02-04 17:47:23 CET
   - removed cruft from package
   - simplified Dist::Zilla handling, moved to Milla
   - expanded interface and added reassemble for "complex" operations

0.21      2015-01-06 19:43:19 Europe/Rome
   - release after 64 passes on CPAN Testers (thanks!)

0.2_03    2015-01-06 07:32:02 Europe/Rome
   - fixed test t/10-writer-add.t for openining a handle to a scalar
     ref in perl prior to 5.14 (no ':raw' allowed, using binmode()
     just after)

0.2_02    2015-01-06 01:38:39 Europe/Rome
   - new D::E::Writer and writer() implementation (sanitized)
   - added documentation for data-embed script
   - documentation bugfixes

0.2_01    2015-01-03 16:41:07 Europe/Rome
   - added package for embedding one file in a Perl module
   - added sub-command in data-embed for feature above

0.1       2015-01-01 09:37:16 Europe/Rome
   - release after 71 passes on CPAN Testers (thanks!)

0.1_02    2014-12-31 14:09:57 Europe/Rome
   - workaround for perl bug in open() prior to version 5.14
   - added name() method to Data::Embed::File
   - added script for embedding/listing/extracting from container files

0.1_01    2014-12-30 17:47:32 Europe/Rome
   - developer release to see what CPAN Testers say


lib/Data/Embed.pm
package Data::Embed;

use strict;
use warnings;
use English qw< -no_match_vars >;
use Exporter qw< import >;
{ our $VERSION = '0.31_1'; }
use Log::Log4perl::Tiny qw< :easy :dead_if_first >;
use Scalar::Util qw< blessed >;

our @EXPORT_OK =
  qw< writer reader embed embedded generate_module_from_file reassemble >;
our @EXPORT      = ();
our %EXPORT_TAGS = (
   all     => \@EXPORT_OK,
   reading => [qw< reader embedded >],
   writing => [qw< writer embed    generate_module_from_file reassemble >],
);

sub writer {
   require Data::Embed::Writer;
   return Data::Embed::Writer->new(@_);
}

sub reader {
   require Data::Embed::Reader;
   return Data::Embed::Reader->new(@_);
}

sub embed {
   my %args = (@_ && ref($_[0])) ? %{$_[0]} : @_;

   my %constructor_args =
     map { $_ => delete $args{$_} } qw< input output >;
   $constructor_args{input} = $constructor_args{output} =
     delete $args{container}
     if exists $args{container};
   my $writer = writer(%constructor_args)
     or LOGCROAK 'could not get the writer object';

   return $writer->add(%args);
} ## end sub embed

sub embedded {
   my $reader = reader(shift)
     or LOGCROAK 'could not get the writer object';
   return $reader->files();
}

sub generate_module_from_file {
   require Data::Embed::OneFileAsModule;
   goto &Data::Embed::OneFileAsModule::generate_module_from_file;
}

sub __compare_and_shift {
   my ($l1, $l2) = @_;
   while (scalar(@$l1) && scalar(@$l2)) {
      last unless $l1->[0]->is_same_as($l2->[0]);
      shift @$l1;
      shift @$l2;
   }
   return ($l1, $l2);
} ## end sub __compare_and_shift

sub __temporary_for {
   my ($target, $previous) = @_;
   require File::Temp;
   require File::Basename;

   my $dir      = File::Basename::dirname $target;
   my $template = File::Basename::basename($target) . '.tmp-XXXXXXX';
   my ($fh, $filename) = File::Temp::tempfile($template, DIR => $dir);
   binmode $fh, ':raw';

   my $prefix = $previous->prefix();
   if ($prefix->{length}) {
      require Data::Embed::Util;
      Data::Embed::Util::transfer($prefix->fh(), $fh);
   }

   close $fh;

   return $filename;
} ## end sub __temporary_for

sub reassemble {
   my %args = (@_ && ref($_[0])) ? %{$_[0]} : @_;

   my $target = $args{target};
   my $interim_target;
   my @sequence = @{$args{sequence} || []};

   my $writer;
   if (ref($target) eq 'SCALAR' || (-e $target)) {
      my $previous = reader($target);

      my ($rprevious, $rsequence) =
        __compare_and_shift([$previous->files()], [@sequence]);

      # is it a nop?
      my $nprevious = scalar @$rprevious;
      my $nsequence = scalar @$rsequence;
      return unless $nprevious || $nsequence;

      # if there's a residual in both, use a temporary something
      if (
         ($nprevious && $nsequence)    # this is the real condition
         || $nprevious    # FIXME move into its own, see next condition...
        )
      {
         if (ref $target) {    # pointer to scalar... hopefully!
            $interim_target = $previous->prefix()->contents();
            $writer         = writer(
               input  => \$interim_target,
               output => \$interim_target,
            );
         } ## end if (ref $target)
         else {
            $interim_target = __temporary_for($target, $previous);
            $writer = writer(
               input  => $interim_target,
               output => $interim_target,
            );
         } ## end else [ if (ref $target) ]
      } ## end if (($nprevious && $nsequence...))
      elsif ($nprevious) {    # we "just" have to get rid of stuff
             # FIXME will implement later, let's just no-reuse here...
      }
      else {    # append residual stuff in @$rsequence
         @sequence = @$rsequence;
         $writer = writer(output => $target, input => $target);
      }
   } ## end if (ref($target) eq 'SCALAR'...)
   else {
      $writer = writer(output => $target);
   }

   $writer->add(
      inputs => [
         map {
            if (blessed($_) && $_->isa('Data::Embed::File')) {
               {
                  name => $_->name(),
                  fh   => $_->fh(),
               };
            } ## end if (blessed($_) && $_->...)
            else {
               $_;
            }
         } @sequence
      ]
   );
   $writer->write_index();

   if (defined $interim_target) {
      if (ref $target) {
         $$target = $interim_target;
      }
      else {
         rename $interim_target, $target;    # atomic
      }
   } ## end if (defined $interim_target)

   return;
} ## end sub reassemble

1;


lib/Data/Embed/File.pm
package Data::Embed::File;

# ABSTRACT: embed arbitrary data in a file

use strict;
use warnings;
use English qw< -no_match_vars >;
use IO::Slice;
use Fcntl qw< :seek >;
use Log::Log4perl::Tiny qw< :easy >;
use Scalar::Util qw< refaddr blessed >;

our $VERSION = '0.30'; # make indexer happy

sub new {
   my $package = shift;
   my $self = {(scalar(@_) && ref($_[0])) ? %{$_[0]} : @_};
   for my $feature (qw< offset length >) {
      LOGCROAK "$package new(): missing required field $feature"
        unless defined($self->{$feature})
        && $self->{$feature} =~ m{\A\d+\z}mxs;
   }
   LOGDIE "$package new(): either filename or fh are required"
     unless defined($self->{fh}) || defined($self->{filename});
   return bless $self, $package;
} ## end sub new

sub fh {
   my $self = shift;
   if (!exists $self->{slicefh}) {
      my %args = map { $_ => $self->{$_} }
        grep { defined $self->{$_} } qw< fh filename offset length >;
      $self->{slicefh} = IO::Slice->new(%args);
   }
   return $self->{slicefh};
} ## end sub fh

sub contents {
   my $self    = shift;
   my $fh      = $self->fh();
   my $current = tell $fh;
   seek $fh, 0, SEEK_SET;

   local $/ = wantarray() ? $/ : undef;
   my @retval = <$fh>;
   seek $fh, $current, SEEK_SET;
   return @retval if wantarray();
   return $retval[0];
} ## end sub contents

sub name { return shift->{name}; }

sub _dname {
   my $name = shift->{name};
   return $name if defined $name;
   return '';
}

sub is_same_as {
   my ($self, $other) = @_;
   return unless blessed($other);
   return unless $other->isa('Data::Embed::File');

   # quick wins
   return unless $self->{offset} == $other->{offset};
   return unless $self->{length} == $other->{length};

   # names must be the same
   return unless $self->_dname() eq $other->_dname();

   # check data sources
   if (defined $self->{fh}) {
      return unless defined $other->{fh};
      return refaddr($self->{fh}) eq refaddr($other->{fh});
   }
   elsif (defined $self->{filename}) {
      return unless defined $other->{filename};    # paranoid...
      if (ref $self->{filename}) {
         return unless ref $other->{filename};
         return refaddr($self->{filename}) eq refaddr($other->{filename});
      }
      return $self->{filename} eq $other->{filename};
   } ## end elsif (defined $self->{filename...})
   else {                                          # paranoid!
      return;
   }

   return 1;                                       # you made it!
} ## end sub is_same_as

1;


lib/Data/Embed/Util.pm
package Data::Embed::Util;

use strict;
use warnings;
use English qw< -no_match_vars >;
use Log::Log4perl::Tiny qw< :easy :dead_if_first >;

our $VERSION = '0.30'; # make indexer happy

use Exporter qw< import >;
our @EXPORT_OK = qw< STARTER TERMINATOR escape unescape transfer >;
our @EXPORT      = ();    # export nothing by default
our %EXPORT_TAGS = (
   all       => \@EXPORT_OK,
   escaping  => [qw< escape unescape >],
   constants => [qw< STARTER TERMINATOR >],
);

use constant STARTER    => "Data::Embed/index/begin\n";
use constant TERMINATOR => "Data::Embed/index/end\n";

sub escape {
   my $text = shift;
   $text =~ s{([^\w.-])}{'%' . sprintf('%02x', ord $1)}egmxs;
   return $text;
}

sub unescape {
   my $text = shift;
   $text =~ s{%(..)}{chr(hex($1))}egmxs;
   return $text;
}

sub transfer {
   my ($infh, $outfh) = @_;
   while ('necessary') {
      my $buffer;
      my $n = sysread $infh, $buffer, 0, 4096;
      LOGCROAK "sysread(): $OS_ERROR" unless defined $n;
      last unless $n;
      print {$outfh} $buffer;
   } ## end while ('necessary')
   return;
} ## end sub transfer

1;


lib/Data/Embed/Writer.pm
package Data::Embed::Writer;

use strict;
use warnings;
use English qw< -no_match_vars >;
use Data::Embed::Reader;
use Log::Log4perl::Tiny qw< :easy :dead_if_first >;
use Data::Embed::Util qw< :constants escape >;
use Fcntl qw< :seek >;
use Scalar::Util qw< refaddr >;

our $VERSION = '0.30'; # make indexer happy

sub __output_for_new {
   my $self    = shift;
   my $package = ref $self;
   my $output  = $self->{output} = $self->{args}{output};
   $self->{output_same_as_input} = 0;    # by default

   # The simple stuff: not present/defined, or the classical "-" string
   if ((!defined($output)) || (!length($output)) || ($output eq '-')) {
      DEBUG $package, "::__output_for_new(): output to STDOUT";
      open my $fh, '>&', \*STDOUT
        or LOGCROAK "dup(): $OS_ERROR";
      binmode $fh
        or LOGCROAK "binmode(\\*STDOUT): $OS_ERROR";
      $self->{output_name} = '<STDOUT>';
      $self->{output_fh}   = $fh;
      $self->{output_type} = 'filehandle';
      return $self;
   } ## end if ((!defined($output)...))

   my $oref = ref $output;
   if (!$oref) {    # filename
      DEBUG $package, '::__output:for_new(): output to a file';
      $self->{output_type} = 'file';
      $self->{output_name} = $output;

      # same file as input? If yes, do not clobber
      if (($self->{input_type} eq 'file') && ($output eq $self->{input})) {
         open my $fh, '+<', $output
           or LOGCROAK "open('$output'): $OS_ERROR";
         binmode $fh
           or LOGCROAK "binmode('$output'): $OS_ERROR";
         $self->{output_fh}            = $fh;
         $self->{output_same_as_input} = 1;
         return $self;
      } ## end if (($self->{input_type...}))

      open my $fh, '>', $output
        or LOGCROAK "open('$output'): $OS_ERROR";
      binmode $fh
        or LOGCROAK "binmode('$output'): $OS_ERROR";
      $self->{output_fh} = $fh;
      return $self;
   } ## end if (!$oref)

   if ($oref eq 'SCALAR') {    # reference to a scalar, similar to filename
      DEBUG $package, '::__output:for_new(): output to a scalar ref';
      $self->{output_type} = 'scalar-ref';
      $self->{output_name} = "{$output}";

      # same file as input? If yes, do not clobber
      if (  ($self->{input_type} eq 'scalar-ref')
         && (refaddr($output) eq refaddr($self->{input})))
      {
         open my $fh, '+<', $output
           or LOGCROAK "open('$output'): $OS_ERROR";
         binmode $fh
           or LOGCROAK "binmode('$output'): $OS_ERROR";
         $self->{output_fh}            = $fh;
         $self->{output_same_as_input} = 1;
         return $self;
      } ## end if (($self->{input_type...}))

      open my $fh, '>', $output
        or LOGCROAK "open('$output'): $OS_ERROR";
      binmode $fh
        or LOGCROAK "binmode('$output'): $OS_ERROR";
      $self->{output_fh} = $fh;
      return $self;
   } ## end if ($oref eq 'SCALAR')

   # Otherwise, we will have to assume that it is a filehandle
   $self->{output_name}          = '<HANDLE>';
   $self->{output_fh}            = $output;
   $self->{output_type}          = 'filehandle';
   $self->{output_same_as_input} = ($self->{input_type} eq 'filehandle')
     && (refaddr($output) eq refaddr($self->{input_fh}));
   return $self;
} ## end sub __output_for_new

sub __input_for_new {
   my $self    = shift;
   my $package = ref $self;
   my $input   = $self->{input} = $self->{args}{input};

   # if not defined, it just does not exist
   if (!defined($input)) {
      DEBUG $package, "::__input_for_new(): no input";
      $self->{input_name} = '*undef*';
      $self->{input_fh}   = undef;
      $self->{input_type} = 'undef';
      return $self;
   } ## end if (!defined($input))

   # the classical "-" string
   if ($input eq '-') {
      DEBUG $package, "::__input_for_new(): input from STDIN";
      open my $fh, '<&', \*STDIN
        or LOGCROAK "dup(): $OS_ERROR";
      binmode $fh
        or LOGCROAK "binmode(\\*STDIN): $OS_ERROR";
      $self->{input_name} = '<STDIN>';
      $self->{input_fh}   = $fh;
      $self->{input_type} = 'filehandle';
      return $self;
   } ## end if ($input eq '-')

   my $iref = ref $input;
   if (!$iref) {    # filename
      DEBUG $package, '::__input:for_new(): input from file';
      open my $fh, '<', $input
        or LOGCROAK "open('$input'): $OS_ERROR";
      binmode $fh
        or LOGCROAK "binmode('$input'): $OS_ERROR";
      $self->{input_name} = $input;
      $self->{input_fh}   = $fh;
      $self->{input_type} = 'file';
      return $self;
   } ## end if (!$iref)

   if ($iref eq 'SCALAR') {    # reference to a scalar, similar to filename
      DEBUG $package, '::__input:for_new(): input from a scalar ref';
      open my $fh, '<', $input
        or LOGCROAK "open('$input'): $OS_ERROR";
      binmode $fh
        or LOGCROAK "binmode('$input'): $OS_ERROR";
      $self->{input_name} = "{$input}";
      $self->{input_fh}   = $fh;
      $self->{input_type} = 'scalar-ref';
      return $self;
   } ## end if ($iref eq 'SCALAR')

   # Otherwise, we will have to assume that it is a filehandle
   $self->{input_name} = '<HANDLE>';
   $self->{input_fh}   = $input;
   $self->{input_type} = 'filehandle';
   return $self;
} ## end sub __input_for_new

sub _transfer_input {
   my $self = shift;
   if (!$self->{input_transferred}) {

      # if there is an input, transfer to the output if it is the case
      if ($self->{input_fh}) {
         if ($self->{output_same_as_input})
         {    # don't copy, assume seekable input
            my $reader = Data::Embed::Reader->new($self->{input_fh});
            my $ifile = $reader->_index();    # private method called
            my @index = $ifile->contents();
            shift @index;                     # eliminate STARTER
            pop @index;                       # eliminate TERMINATOR
            $self->{index} = \@index;    # initialize previous index
                                         # put out handle in right position
            seek $self->{output_fh}, $ifile->{offset}, SEEK_SET;
         } ## end if ($self->{output_same_as_input...})
         else {
            my $starter    = STARTER;
            my $terminator = TERMINATOR;
            my (@index, $index_completed);
            my $ifh = $self->{input_fh};
            my $ofh = $self->{output_fh};
          INPUT:
            while (<$ifh>) {
               if (!@index) {
                  if ($_ eq $starter) {
                     push @index, $_;
                     next INPUT;
                  }
                  else {
                     print {$ofh} $_;
                  }
               } ## end if (!@index)
               elsif (!$index_completed) {    # accumulating index
                  if (m{\A \s* (\d+) \s+ (\S*) \s*\z}mxs) {
                     push @index, $_;
                  }
                  elsif ($_ eq $terminator) {
                     push @index, $_;
                     $index_completed = 1;
                  }
                  else {    # not a valid index, flush accumulated lines
                     print {$ofh} @index;
                     @index           = ();
                     $index_completed = undef;    # paranoid
                  }
               } ## end elsif (!$index_completed)
               else
               {    # accumulating and index completed, but other stuff...
                  print {$ofh} @index;    # flush and reset
                  @index           = ();
                  $index_completed = undef;
               } ## end else [ if (!@index) ]
            } ## end INPUT: while (<$ifh>)
            shift @index;                 # eliminate STARTER
            pop @index;                   # eliminate TERMINATOR
            $self->{index} = \@index;     # initialize previous index
         } ## end else [ if ($self->{output_same_as_input...})]
      } ## end if ($self->{input_fh})
      $self->{input_transferred} = 1;
   } ## end if (!$self->{input_transferred...})
   return $self;
} ## end sub _transfer_input

sub new {
   my $package = shift;
   my %args = (scalar(@_) && ref($_[0])) ? %{$_[0]} : @_;
   $args{transfer_input} = 1 unless exists $args{transfer_input};

   # Undocumented, keep additional parameters around...
   my $self = bless {args => \%args}, $package;

   # first of all, resolve the input
   $self->__input_for_new();

   # then the output (might depend on the input)
   $self->__output_for_new();

   $self->_transfer_input() if $args{transfer_input};

   return $self;
} ## end sub new

sub add {
   my $self = shift;
   my %args = (scalar(@_) && ref($_[0])) ? %{$_[0]} : @_;

   $self->_add($_)
     for (defined($args{inputs}) ? @{$args{inputs}} : (\%args));

   return $self;
} ## end sub add

sub _add {
   my ($self, $args) = @_;

   # DWIM!
   if (defined $args->{input}) {
      if ($args->{input} eq '-') {
         open my $fh, '<&', \*STDIN
           or LOGCROAK "dup(): $OS_ERROR";
         binmode $fh
           or LOGCROAK "binmode(\\*STDIN): $OS_ERROR";
         $args->{fh} = $fh;
      } ## end if ($args->{input} eq ...)
      else {
         my $ref = ref $args->{input};
         if ((!$ref) || ($ref eq 'SCALAR')) {
            $args->{filename} = $args->{input};
         }
         else {
            $args->{fh} = $args->{input};
         }
      } ## end else [ if ($args->{input} eq ...)]
   } ## end if (defined $args->{input...})

   if (defined $args->{fh}) {
      return $self->add_fh(@{$args}{qw< name fh >});
   }
   elsif (defined $args->{filename}) {
      return $self->add_file(@{$args}{qw< name filename >});
   }
   elsif (defined $args->{data}) {
      return $self->add_data(@{$args}{qw< name data >});
   }

   LOGCROAK "add() needs some input";
   return;    # unreached
} ## end sub _add

sub add_file {
   my ($self, $name, $filename) = @_;
   $name = '' unless defined $name;
   my $print_name =
     (ref($filename) eq 'SCALAR') ? 'internal data' : $filename;
   DEBUG "add_file(): $name => $filename";

   # To make it work with references to scalars in perl pre-5.14
   # we split open() and binmode()
   open my $fh, '<', $filename
     or LOGCROAK "open('$print_name'): $OS_ERROR";
   binmode $fh
     or LOGCROAK "binmode('$print_name') failed";

   return $self->add_fh($name, $fh);
} ## end sub add_file

sub add_data {
   my ($self, $name) = @_;
   return $self->add_file($name, \$_[2]);
}

sub add_fh {
   my ($self, $name, $input_fh) = @_;

   my $output_fh   = $self->{output_fh};
   my $data_length = 0;                    # go!
   while (!eof $input_fh) {
      my $buffer;
      defined(my $nread = read $input_fh, $buffer, 4096)
        or LOGCROAK "read(): $OS_ERROR";
      last unless $nread;                  # safe side?
      print {$output_fh} $buffer
        or LOGCROAK "print(): $OS_ERROR";
      $data_length += $nread;
   } ## end while (!eof $input_fh)

   # Add separator, not really needed but might come handy for
   # peeking into the file
   print {$output_fh} "\n\n"
     or LOGCROAK "print(): $OS_ERROR";

   $name = '' unless defined $name;
   push @{$self->{index}}, sprintf "%d %s\n", $data_length, escape($name);

   return $self;
} ## end sub add_fh

sub write_index {
   my $self = shift;
   my ($output_fh, $index) = @{$self}{qw< output_fh index >};
   print {$output_fh} STARTER, @$index, TERMINATOR
     or LOGCROAK "print(): $OS_ERROR";
   delete $self->{$_} for qw< output_fh index >;
   return;
} ## end sub write_index

sub DESTROY {
   my $self = shift;
   $self->write_index() if exists $self->{output_fh};
   return;
}

1;
